### ISSUE TITLE
Major DataBase overhaul and restructuring [$20]

### ISSUE THREAD

**Issue by E3V3A** [CellularPrivacy/Android-IMSI-Catcher-Detector#215]
As we have noticed in many recent issues and many past discussions, there is a terrible urgent need to restructure our original database (DB). This DB was originally made for our very first attempts in building this app. However, we've come to realize that the way it is used and structured and how its tables are internally labelled is really confusing at best. It was made according to very weak documentation and poor understanding of what we actually needed. 

After several weeks of thinking about this, on and off, I've come to the point where it's time to get it done. Here is a brand new proposal that should cover most features in the near future and hopefully most of the **beta** stage life-time. 

This is the way we will do it:
 1. I finish the DB documents and table ER diagram, then 
 2. you guys make a critical re-view and feedback.
 3. I make the required updates, if any, and go back to (1), else 
 4. we re-code the DB into app.

**NOTE**: I suggest keeping the old tables while adding the new ones, to maintain simple and progressive migration. 

There will surely be loads of questions surrounding this, but I think I might have covered most of it in the (still to be finished) documents:
 - [x] General DB functional overview
 - [ ] Detailed DB tables Descriptions
 - [ ] Default DB data 

Then there will be new App help pages and an FAQ, to be implemented inside app. (Separate issue.)
 - [x] General App functionality
 - [x] Internal App help pages
 - [ ] Internal App FAQ
 - [ ] *TBA*

---

#### Here are the **old** currently used DB tables:

![aimsicd_mycellinfo_er_2](https://spideroak.com/share/IFEU2U2JINCA/GitHub/home/SecUpwN/SpiderOak/DOCUMENTATION/Database_myCellInfo_ER_2.png)

---

#### Here are the **NEW** DB tables:
**`UPDATED: 2015-07-31`**

![aimsicd5_er](https://spideroak.com/share/IFEU2U2JINCA/GitHub/home/SecUpwN/SpiderOak/DOCUMENTATION/Database_myCellInfo_ER_3.png)

---

**Reference Issues:** These are issues that directly depend on these tables. 
#13, #69,#71, #91, #93, #97, #100, #117, #176, #201, (#203 ?), 

More descriptions will be added below.

**Comment by SecUpwN** [comment]
@tobykurien, here ya go, now is our chance to rock this stuff since @E3V3A will be away for a week. so let's surpise him, shall we? please tell me what exactly I can tackle here and I'll do it right away!

**Comment by E3V3A** [comment]
Updated the *new* ER-diagram with an `EventLog` table, to be used to export events. Since the event log should be persistent even after wiping the `DBi_*` and `DBe_*` tables, it is put in its own table, instead of using SQL connections.

**Comment by SecUpwN** [comment]
Just found [greenDAO](http://greendao-orm.com/) by @greenrobot, which is essentially an open source project [on GitHub](https://github.com/greenrobot/greenDAO) to help Android developers working with data stored in SQLite. Thought I'd toss that in.

**Comment by E3V3A** [comment]
@SecUpwN Thanks for trying to help, but this is already covered. We a;already have full access to any DB SW tools we need. (Do you remember adding those sponsor images?)

**Comment by SecUpwN** [comment]
>Do you remember adding those sponsor images?

Oh right. Thank you for reminding me. I'm outa here for now.

**Comment by E3V3A** [comment]
I have an SQL script `aimsicd.sql`, that will construct the entire schema of the new tables above. This need to be distributed by our app and then imported from command line (via Java call) with:
```
# cd /data/data/com.SecUpwN.AIMSICD/databases/
# cat aimsicd.sql | sqlite3 aimsicd.db
```
Where shall I put it? I.e. in what GitHub directory? in [assets](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/tree/master/app/src/main/assets)?

**Comment by SecUpwN** [comment]
`assets/databases` should be the right place. Also found [Android SQLiteAssetHelper](https://github.com/jgilfelt/android-sqlite-asset-helper) by @jgilfelt.

**Comment by E3V3A** [comment]
@SecUpwN Good thing to know, but I think we've got this one covered. We don't need dozens of pages of code to do this. Just the two lines above. We don't even need those lines, if we supply a pre-populated `aimsicd.db`. That saves a lot of Java coding!

**Comment by E3V3A** [comment]
We have to implement a minimal part of the new Database Table structure so that we can move on to the new **Beta** phase of detection development. Most of our subsequent detection mechanisms depend on the two table blocks "DBi" and "DBe".

The easiest way to do this is by:

 1. Add the tables called `DBi_measure` and `DBi_bts`, which are connected by the **bts_id** column. However, for simplicity we can start by just putting the `DBi_bts` columns into the `DBi_measure` table, although I'd prefer to do it right from the beginning. 

 2. Change the column names in the old `opencellid` table to those found in `DBe_external`. Then change the table name, once working.

 3. The tricky part is to migrate the mixed up table items of `cellinfo` and `locationinfo` into the `DBi_measure` (and later also the `DBi_bts`) table. This mix-up is the origin of where most confusions occurs.

 4. All other tables are either already there and only need minor column name/type adjustments or are completely new and independent.

**Comment by E3V3A** [comment]
@tobykurien  What is the purpose and use of the `locationinfo`:*Connection* column? 
(Where is it used and for what?)

**Comment by tobykurien** [comment]
It appears to be an informational column, not used anywhere.

**Comment by indoraq** [comment]
I suggest two separate measurement tables:
1. GPS measurements (GPS-fix facts)
2. CGI  measurements (RAN on NW-fix facts)

The point is that these two measurements on independant. You can have GPS-fix without any CGI and vice-versa: no GPS but CGI-s. First can happen in rural, last is quite common in urban settings: if one moves in a shopping center then no GPS-fix, but there can be a lot of CGI-s.

**Comment by E3V3A** [comment]
We're not interested in the first scenario, so there's no need for just saving gps points without network.  Also as measurements are taken, and there is no gps data available, we should just leave that field empty for that measurement, and not copying last location.

**Comment by indoraq** [comment]
You can't say that. For NW-fix there are situations when the CGI is either incorrect (invalid) or can't be obtained for whatever reasons. It is not about areas without RAN coverage but happens in urban situations when UE is attached to RAN and services are available. In data model you can overcome this by applying dummy values, but the semantic ambiguity  of timestamp can not be avoided by this.

**Comment by E3V3A** [comment]
@indoraq Please send me an email (put yours in the body) and I'll send you link to our internal chat and we can take things further from there.

**Comment by indoraq** [comment]
Indrek Pehk Indrek.Pehk@reach-u.com<mailto:Indrek.Pehk@reach-u.com>

**Comment by SecUpwN** [comment]
@E3V3A, are you continuing to work on this, or did you drop us?

**Comment by E3V3A** [comment]
I think we need a developer update on what tables are currently present. (I know @banjaxbanjo added some.) Then we need to add the code to build the tables as defined above (+ additions) and we can start moving and changing the code from using the old tables to the new ones. Most *template*-like structure is already in there and it should be fairly easy to copy/paste the rest.

**Comment by E3V3A** [comment]
I'm adding the "bug" label, and removing the "dev_discussion" label, so that new developers will understand that this is a real issue. 

In the meantime, we can try to merge the data from `Measured Signal Strengths` table into the `BTS Measurements` table (as visible from the *Database Viewer*.) This might need to be posted as a separate issue. I also noticed that the signal strengths (RSS) as shown in the former is no longer working...

**Comment by E3V3A** [comment]
I'm adding the "bug" label, and removing the "dev_discussion" label, so that new developers will understand that this is a real issue. 

In the meantime, we can try to merge the data from `Measured Signal Strengths` table into the `BTS Measurements` table (as visible from the *Database Viewer*.) This might need to be posted as a separate issue. I also noticed that the signal strengths (RSS) as shown in the former is no longer working...

**Comment by banjaxbanjo** [comment]
Creating the tables via java code is not to difficult
https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/blob/development/app/src/main/java/com/SecUpwN/AIMSICD/smsdetection/SmsDetectionDbHelper.java

but is going to be a little work creating the insert,update and delete code for the tables and tracing the old database code in the project since  its so intwined in the app.

Also will users lose all there own data if a major overhaul of database is done since its structure is alot different.

**Comment by SecUpwN** [comment]
>but is going to be a little work creating the insert,update and delete code for the tables and tracing the old database code in the project since its so intwined in the app.

Would you be able to tackle this, possibly in smaller steps?

>Also will users lose all there own data if a major overhaul of database is done since its structure is alot different.

Don't worry about this. Our app displays `ALPHA` everywhere, so we have every right too change it.

**Comment by E3V3A** [comment]
> Creating the tables via java code is not too difficult

@banjaxbanjo  The code as linked above is just fine, but in general we're **not** creating DB code by using java code to populate an SQL based DB. 

Generally speaking, using the java Google API for table manipulation, Is completely senseless since it is extremely non-transparent. SQL was created for simplicity, portability and ease to understand, so we're trying to keep the SQL statements as close to native as possible. This allows for easy export, backup, DB debugging and maintenance. Which is the reason we have [aimsicd.sql](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/blob/d7a9f64e3f4c1e288b0288299d32313273744d22/app/src/main/assets/aimsicd.sql). In addition, there are already many placeholders for copy/paste implementation of what we already have, to support the new tables. Just remove the commented out code, and rename the tables as described in the comments.

**Comment by SecUpwN** [comment]
>...refusing to implement #215...

We're *not* "refusing" to implement the changes to fix this Issue as you've just said in https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/471#issuecomment-112507019. We still lack developers and apparently this sort of behavior keeps pissing off people. Let's *not demand*, but kindly ask and *contribute bounties* like I did a few days ago here to keep them motivated. @marvinmarnhold, @wvengen and @DJaeger, would you please have a look at the previous post above mine to see if you can tackle this? Our future development as well as fixing numerous other bugs heavily relies on implementing the changes in this Issue correctly. I hope that the bounty for this Issue will be raised further so that everyone contributing pull requests will get a fair share. Cheers!

**Comment by E3V3A** [comment]
@SecUpwN  As you can see this issue is more than 7 months old, and has been considered top priority ever since. Yet, nobody want to take care or even begin on this issue, and rather start trying to implement new DB tables for non/poorly-documented features or modify the old ones, breaking loads of functional dependency in the meantime. I'm sick of seeing this issue still open. **It's blocking all future detection development in #230.**

**Comment by marvinmarnold** [comment]
I can definitely try to get this done but I'm swamped through the end of the month. Can it wait until early July?

**Comment by SecUpwN** [comment]
>I can definitely try to get this done but I'm swamped through the end of the month.

Smaller steps are better than no steps at all. If you can do literally *anything*, that'd be great! ;-)

>Can it wait until early July?

Possibly yes. Of course this Issue is burning heavily under our asses as it introduces more and more Issues that would likely not even appear if we'd finally get this done as quickly as possible.

**Comment by marvinmarnold** [comment]
Gotcha. I'll try to take some steps in the interim... but no promises till July.

**Comment by SecUpwN** [comment]
>Gotcha. I'll try to take some steps in the interim...

That'll work for us. Move as slow as you wish.

>but no promises till July.

Please never promise anything. Just do. ;-)

**Comment by DJaeger** [comment]
Will look into

**Comment by E3V3A** [comment]
A good start for this work is my [comment](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/215#issuecomment-71375823) and in the meantime I'll try to update the reference in [aimsicd.sql](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/blob/259b169048d09943706e6796a5dc1e80af994b56/app/src/main/assets/aimsicd.sql). (NB. This file is currently not used anywhere, but I use it as a reference model while creating the diagrams.)

**Comment by SecUpwN** [comment]
@E3V3A, @banjaxbanjo and @DJaeger, it seems @agilob has had a good idea in https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/486#issuecomment-114561916:

>Database storage should be rewritten using ORM, none of those problems would exist ;_;

**Comment by wvengen** [comment]
At the very least, some database [migration](http://stackoverflow.com/questions/2488813/how-to-do-database-schema-migrations-in-android) system would be useful. As it looks, I'd say an ORM could be a good choice - [one of these](http://www.appbrain.com/stats/libraries/tag/database/database-libraries) or [more](http://stackoverflow.com/questions/371538/any-good-orm-tools-for-android-development).

**Comment by E3V3A** [comment]
I'm not gonna have this silly ORM discussion again! (Please search and re-read ALL the DB threads!) 
This is not about ORM, but about implementing the bloody DB design as specified and not wasting time in complicating the code further. Learn SQL or stop programming. If you don't know how SQLite (SQL) works, you sure as hell not gonna be able to do anything useful with ORM.

**Comment by okbob** [comment]
For first quick view it is looking well. Every schema should be validated on prototype. Insert to tables 100 rows, and try to write queries for all planned processes. When the queries are readable and simple, then the design is perfect. If not, try to fix it. Usually it needs two three iterations. Next step should be a performance validation - try to fill database by data for 5 years, and check a performance of prepared queries.

**Comment by E3V3A** [comment]
We need to seriously consider how treat multiple-threads that are causing locked issues like in #486 and #528. The way to go seem to be to use a single connection and keep it open for the duration of the app life. The best references  I could find are these: 

[Single SQLite connection](http://touchlabblog.tumblr.com/post/24474750219/single-sqlite-connection)
[Android Sqlite Locking](http://touchlabblog.tumblr.com/post/24474398246/android-sqlite-locking)
[Android SQLiteDatabase locking and multi-threading](https://nfrolov.wordpress.com/2014/08/16/android-sqlitedatabase-locking-and-multi-threading)


[stackoverflow: android-sqlite-with-multi-thread](http://stackoverflow.com/questions/10450977/android-sqlite-with-multi-thread)
[stackoverflow: what-are-the-best-practices-for-sqlite-on-android](http://stackoverflow.com/questions/2493331/what-are-the-best-practices-for-sqlite-on-android/3689883#3689883)
[stackoverflow: sqlite-db-accessed-from-multiple-threads](http://stackoverflow.com/questions/6225374/sqlite-db-accessed-from-multiple-threads)
[stackoverflow: android-sqlite-multithread-access](http://stackoverflow.com/questions/12974219/android-sqlite-multithread-access)

**Comment by marvinmarnold** [comment]
Has anyone started to work on this entire bug or should I take ownership?

**Comment by banjaxbanjo** [comment]
@marvinmarnold I'm currently working on #215 but will probably look into this while in the process of updating db I will email you my current dev what I have so far.

**Comment by marvinmarnold** [comment]
@banjaxbanjo great i made my email public in my github profile

**Comment by banjaxbanjo** [comment]
@marvinmarnold haha silly me didn't even think about looking there

**Comment by SecUpwN** [comment]
@banjaxbanjo, just found [Android DbInspector](https://github.com/infinum/android_dbinspector), may be of interest to you.

**Comment by E3V3A** [comment]
Let us not distract... good work in progress please.

**Comment by SecUpwN** [comment]
>Let us not distract... good work in progress please. 

Sorry for interrupting again, maybe this makes @banjaxbanjo's work easier:

[![Flyway - Database Migrations Made Easy](http://flywaydb.org/assets/logo/flyway-logo-tm.png "Flyway")](https://github.com/flyway/flyway)

**Comment by E3V3A** [comment]
**Congratulations!!**
The Latest merged PR 22c0ca07599b0d62ec7816f3fcce27bfe33da9d6 essentially fixes this. Now we need to do a shit-load of testing and polishing, before closing this.

**Comment by E3V3A** [comment]
@SecUpwN  We need to close this and pay out the bunty, please give relevant people instructions. I'll let you close this one, so not to mess up the bounty.

**Comment by SecUpwN** [comment]
>We need to close this and pay out the bunty, please give relevant people instructions.

This will be done automatically via Bountysource, no worries.

What about this? Shall I still close this Issue as **fixed**? ;-)

>Now we need to do a shit-load of testing and polishing, before closing this.

**Comment by DJaeger** [comment]
I think this should be done before, as this issue is not completely solves before.
I would only want to pay for a complete and not only a half work.

**Comment by SecUpwN** [comment]
>I think this should be done before, as this issue is not completely solved before.
>I would only want to pay for a complete and not only a half work.

Indeed, I agree with you. Bountysource asks donators to agree if an Issue is really fixed before paying out the bounty. In this case, @agilob is the only one who donated, thus he will be asked when closed.

@E3V3A, please see you OP and tick the boxes that have been solved so that they appear checked when looking at the Issue without opening it. Also, please exactly state below how we ought to test this!

**Comment by banjaxbanjo** [comment]
@DJaeger 

Lol
> I would only want to pay for a complete and not only a half work.

$10 for this much coding a the bargin  of the century☺

**Comment by SecUpwN** [comment]
@E3V3A, I have just sent @banjaxbanjo a EUR 50 UKASH code since he really did an awesome job. Please take some quality time to add a testing procedure here so that we can work to finally close this.

**Comment by SecUpwN** [comment]
>Please take some quality time to add a testing procedure here so that we can work to finally close this.

@E3V3A, why won't you respond on this important question? Again "no time and on the phone"? ;-)
@DJaeger, if you've got a good testing procedure in mind, feel invited to post your proposal here.

**Comment by E3V3A** [comment]
> why won't you respond on this important question?

Because every time I ask someone to test something, there is no response. So I'm done with writing testing procedures that nobody give a fcuk about. Some people don't even bother to test their own PR's...

**Comment by SecUpwN** [comment]
>So I'm done with writing testing procedures that nobody give a fcuk about.

@E3V3A, I can understand your frustration and will work on enforcing people to respond to your questions if not answered within a few days in the future. Please add a testing procedure for this now.

**Comment by E3V3A** [comment]
I've just updated OP with a new ER design image. It is based on the **latest** [aimsicd5.sql](https://github.com/E3V3A/woork/blob/f9babf59764b52b0e522669be8472da487a51550/aimsicd5.sql) SQL script, that we should eventually use to create our original **aimsicd.db** file. However, the migration is not yet fully complete, as there were many small changes after code migration. These now need to be fixed. And developers are required to follow the ER diagram for fixing issues and implementing new functions. As **I'm the maintainer of the DB create script**, please contact me if you have any special design requirements.

@SecUpwN can you please copy the image in OP and place it on the wiki where it belongs?

**Comment by E3V3A** [comment]
**Addendum:**
The new "con_state" column in `DBi_measure` should contain the AOS current connection and 
data activity states as given by the API with:
```
  DATA_ACTIVITY:   No,In,Ou,IO,Do
  DATA_CONNECTION: Di,Ct,Cd,Su
  CALL_STATE:      idle,ringing,offhook
  SERVICE_STATE:   emergency_only,in,out,off
```
See: [PhoneStateListener](http://developer.android.com/reference/android/telephony/PhoneStateListener.html) developer info. 

We should also add the current *roaming state* to this column..

**Comment by SecUpwN** [comment]
>@SecUpwN can you please copy the image in OP and place it on the wiki where it belongs?

@E3V3A, done and made available in [Technical Overview](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/wiki/Technical-Overview).

**Comment by E3V3A** [comment]
Addendum:
Because signals measured in [dBm] are always negative and signals measured in [ASU] (Arbitrary Strength Unit) are can be both positive and negative in UMTS, we should change these fields to use `REAL` data type. 
 - `DBe_import:avg_signal`: **REAL**
 - `DBi_measure:rx_signal`: **REAL**

**Comment by SecUpwN** [comment]
Ok, so we've got to finally solve this Issue. @He3556 and @DJaeger, would you please re-read this Issue and find out how to add the missing pieces? I am willing to raise the bounty for this if you guys do!

**Comment by SecUpwN** [comment]
Good morning members of @CellularPrivacy/database! This Issue is one of the most important Issues we need to solve ASAP. It has been open for nearly 2 years, was one of the hardcore reasons why E:V:A (our project inventor) left and *still* blocks all further serious IMSI-Catcher detection mechanisms. I would like **everyone** from the database team to closely read this whole Issue (yes, I know it's a pain, but necessary) so that we can move forward to finally close this bug for good. Keep in mind: *Very* **URGENT**!

**Quick sum-up of what you need to have a look at:**

* Read the [functional table overview](https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/issues/215#issuecomment-67258750). @banjaxbanjo, which tables are still missing?
* Update **con_state** column to contain [AOS current connection and
data activity states](https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/issues/215#issuecomment-126779418)
* Change `DBe_import:avg_signal` and `DBi_measure:rx_signal` to [REAL data type](https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/issues/215#issuecomment-127074842)

In the end, this is what our database needs to look like (documented in [Technical Overview](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/wiki/Technical-Overview)):

![aimsicd5_er](https://spideroak.com/share/IFEU2U2JINCA/GitHub/home/SecUpwN/SpiderOak/DOCUMENTATION/Database_myCellInfo_ER_3.png)

**Comment by larsgrefer** [comment]
I've started the branch [feature/realm](https://github.com/CellularPrivacy/Android-IMSI-Catcher-Detector/tree/feature/realm)

PR's to this branch are welcome

**Comment by SecUpwN** [comment]
>I've started the branch feature/realm

Thanks, @larsgrefer. I hope this brings us forward. A few questions the others might have as well:

1. Will `REALM` be able to cover all the planned tables?
2. What are the advantages over the current SQL table?
3. Have you read this whole thread and all questions?

Our originial inventor E:V:A (who officially left our project and now mainly idles and monitors what we do) was mostly against anything else than SQL and did make very valid points about that. Since he was (is) our database guru, I feel like it is a good thing mentioning him here in the hope that you guys will have a friendly and respectful discussion here. @E3V3A, I would be happy to see your questions to @larsgrefer.

**Comment by larsgrefer** [comment]
> How come you specifically decided to move to REALM?

It seems very interesting to me and I'd like to give it a try

> Will REALM be able to cover all the planned tables?

Sure

> What are the advantages over the current SQL table?

We don't have to worry about OR-Mapping and get a nice java-api.
And of course we can get rid of our 2000+ lines monster-class

**Comment by Nordlenning** [comment]
20 USD bounty on this issue so far
https://www.bountysource.com/issues/7069218-major-database-overhaul-and-restructuring

**Comment by SecUpwN** [comment]
Manually updated this thread title as well and would like to reward it to @larsgrefer when merging #827.

**Comment by larsgrefer** [comment]
resolved for now in #827 and #655

**Comment by agilob** [comment]
Not resolved. This issue was about expanding data categories that are collected by AIMSICD to get more information about cells, networks and devices! Not about using ORM as database middle-man!

**Comment by larsgrefer** [comment]
Many of the problems/whishes named in this issue are solved.

Our persistence layer is on a clean state now. We have nicely named classes and fields. There are no unused classes or fields. The multi-threading problem is also fixed.

**Comment by SecUpwN** [comment]
I agree with you. The main move to a working and easily expandable database has been accomplished.

@larsgrefer, even if not much, I'd like to say **THANK YOU** for investing heart blood with the $20 bounty.
Since @Nordlenning has been a [donator for this Issue](https://www.bountysource.com/issues/7069218-major-database-overhaul-and-restructuring/backers) as well, I'd like him to state if he agrees.

**Comment by Nordlenning** [comment]
Moving to a working and easily expandable database is great.
I would like to reward it to @larsgrefer  His energy is needed elsewhere..
